### PR TITLE
Expose the witnesses of the tx being hashed in SigHashCache

### PR DESCRIPTION
**Pull request by TheBlueMatt**
See docuemntation of the new method for more, but this allows
certain use patterns which were broken with the introduction of
SigHashCache.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Note that this appears to break rust's auto-& stuff, though I don't think that's a big deal. It would imply a major version bump against current SigHashCache, but I don't think we've released it yet, so its fine.

**Comment by TheBlueMatt** [comment]
@instagibbs mind taking a look if this works for you? Given its your change.

File: src/util/bip143.rs [hunk 1]
```diff
@@ -102,9 +104,9 @@ impl SighashComponents {
 }
 
 /// A replacement for SigHashComponents which supports all sighash modes
-pub struct SigHashCache<'a> {
+pub struct SigHashCache<R: Deref<Target=Transaction>> {
```
Comment (dpc) [inline src/util/bip143.rs:107]: I was initially concerned if this is an unidiomatic abuse of `Deref`, but after some thought I think it's fine. `SigHashCache` works with any smart pointer to `Transaction` which is what `Deref` is for.

**Review by jrawsthorne — APPROVED** [review]
LGTM. Any reason for the extra scope in signature_hash()?

**Comment by TheBlueMatt** [comment]
Without the new scope borrowck gets mad that we're holding a ref to self.tx because self.hash_outputs() requires &mut self.

**Comment by jrawsthorne** [comment]
Ah yeah without the newer borrow checker, another reason to upgrade to a newer rust version

File: src/util/bip143.rs [hunk 2]
```diff
@@ -215,6 +219,32 @@ impl<'a> SigHashCache<'a> {
     }
 }
 
+impl<R: DerefMut<Target=Transaction>> SigHashCache<R> {
+    /// When the SigHashCache is initialized with a mutable reference to a transaction instead of a
```
Comment (instagibbs) [inline src/util/bip143.rs:223]: only thought I have is if you could move this comment into a real code block in unit test?

Comment (instagibbs) [reply src/util/bip143.rs:223]: otherwise no objection

Comment (TheBlueMatt) [reply src/util/bip143.rs:223]: Rust is magic. As written that block is run in tests.

**Comment by instagibbs** [comment]
Does rust weigh the same as a 🦆 by any chance

**Review by sgeisler — APPROVED** [review]

File: src/util/bip143.rs (hunk 2 above)
Comment (dpc) [reply src/util/bip143.rs:223]: And you don't have to look around to find this test now. Doctests are wonderful.

**Review by apoelstra — APPROVED** [review]

**Pull request merged**